### ISSUE TITLE
vmod-querystring enterprise edition

### ISSUE THREAD

**Issue by Dridi** [Dridi/libvmod-querystring#23]
## Background

This is an approximately 3-years old idea (precisely from around Apr 25th 2013) of taking advantage of a new feature for an unreleased Varnish 4. Namely VMOD objects.

Having supported this VMOD for Varnish 3 until its EOL, I've never done it and in the mean time Varnish 4.1 was released with more cool stuff. So this time I've decided to lower support for Varnish 4.0 down to bug fixes only and do a 4.1-only rewrite of the module.

You can try it in its dedicated branch:
https://github.com/Dridi/libvmod-querystring/tree/feature/obj

Documentation should be up-to-date but I'll describe the main difference here. YMMV and all changes may not appeal to everyone, but I think it's the best compromise. It has more capabilities with a reduced code base and a lower footprint.

## Object initialization

The big change is the introduction of VMOD objects instead of plain functions. It has the inconvenience of adding one indirection and the associated overhead.

```perl
# before
sub vcl_recv {
    # remove google analytics query params
    set req.url = querystring.regfilter(req.url, "^utm_.*");
}

# after
sub vcl_init {
    # a filter for google analytics query params
    new ga = querystring.filter();
    ga.add_regex("^utm_.*");
}

sub vcl_recv {
    set req.url = ga.apply(req.url);
}
```

There are however some benefits that deserve to be mentioned. Filters have names, which may improve readability and of something like `ga` is too short you may pick a less abbreviated name like `googanal` or `analytics`. Also if your regex is not valid, it will fail the `vcl.load` operation instead of logging it every time it is used (who reads the logs anyway?) which also means that the regex is compiled only once.

This looks also better when using an exact-match criteria, you no longer need to clutter your VCL code with `querystring.filtersep()` which is AFAIK the longest short separator ever. Did you notice how separator is abbreviated to just sep? :trollface:

## Composition

If you need to apply more than one filter, and on top of that sort the remaining parameters, you can now do that with a single filter.

```perl
# before
sub vcl_recv {
    # removed query params:
    # - google analytics
    # - timestamp-based cache bypass
    # - anti-CSRF token
    set req.url = querystring.globfilter(req.url, "utm_*");
    set req.url = querystring.filter(req.url, "_");
    set req.url = querystring.regfilter(req.url, "sess[0-9]+");

    set req.url = querystring.sort(req.url);
}

# after
sub vcl_init {
    new qf = querystring.filter(sort = true);
    qf.add_glob("utm_*"); # google analytics parameters
    qf.add_string("_"); # timestamp-based cache bypass
    qf.add_regex("sess[0-9]+"); # anti-CSRF token
}

sub vcl_recv {
    set req.url = qf.apply(req.url);
}
```

Of course the "before" example is exaggerated and could be solved with a single call to `regfilter` (+ `sort`) but the point is that the "after" example handles it better and requires a lot less workspace to do so. Enabling sorting in a filter has no space overhead whatsoever, only time.

## Better semantics

What would be the `*filter` without their `*filter_except` counterparts? With objects it can be solved in different places like the constructor, the `add_*` methods, or the `apply` method. I opted for the latter, even though it's not the most flexible of all, it's the best compromise wrt simplicity.

There's no `apply_except` function but instead an optional mode `ENUM` (like the optional sort `BOOL` in the constructor, 4.1-only).

```perl
# before
sub vcl_recv {
    set req.url = querystring.filter_except(req.url, "q");
}

# after
sub vcl_init {
    new search = querystring.filter();
    search.add_string("q");
}

sub vcl_recv {
    set req.url = search.apply(req.url, mode = keep);
    # or just
    set req.url = search.apply(req.url, keep);
}
```

I find it a lot easier to reason about negating filters, especially for regular expressions.

## Misc

The `clean`, `sort` and `remove` shorthand functions are still here and rely on filters under the hood.

There's also a new `.extract` method similar to `.apply` except that it discards the rest of the URL and only returns the query-string.

The code style evolved from my 2012 self's to Varnish's, making it more maintainable. There's also less code to maintain but more documentation now.

## Feedback

This work is already done and testable, including documentation. I still have to cover some cases in the test suite, so until I'm done I'm interested in your feedback.

Am I missing something obvious?
Do you have comments?

Thanks,
Dridi

**Comment by erickfaustino** [comment]
Hello, @Dridi. Sorry for my late reply, but I'm in a hospital due a motorcycle crash, no worries, I'm fine.
For me, it's look more flexible and readable, although I don't know nothing about C (I'm a just system administrator)
My feedback wouldn't be so helpful at the developer's side, but I'm glad that you are maintaining and improving this great VMOD. 
Thanks!
Erick.

**Comment by Dridi** [comment]
It's been released in 1.0 today!
